File: src/domain/recommendation.ts
import type { Recommendation, RentalListing, RouteEstimate, VehicleMatch } from "./types";

const MATCH_PRIORITY: Record<VehicleMatch["kind"], number> = {
  exact: 3,
  "similar-class": 2,
  "low-confidence": 1
};

export function buildRecommendation(
  listing: RentalListing,
  match: VehicleMatch,
  taxiRoute: RouteEstimate,
  transitRoute: RouteEstimate
): Recommendation {
  const rentalTotal = roundMoney(
    listing.basePrice + listing.platformFees + listing.insuranceFees + listing.oneWayFee
  );
  const taxiTotal = roundMoney(rentalTotal + taxiRoute.cost);
  const transitTotal = roundMoney(rentalTotal + transitRoute.cost);
  const bestRouteMode = taxiTotal <= transitTotal ? "taxi" : "transit";

  return {
    listing,
    match,
    taxiRoute,
    transitRoute,
    rentalTotal,
    taxiTotal,
    transitTotal,
    bestTotal: bestRouteMode === "taxi" ? taxiTotal : transitTotal,
    bestRouteMode,
    warnings: Array.from(new Set(listing.warnings))
  };
}

export function rankRecommendations(recommendations: Recommendation[]): Recommendation[] {
  return [...recommendations].sort((left, right) => {
    const totalDelta = left.bestTotal - right.bestTotal;
    if (totalDelta !== 0) {
      return totalDelta;
    }

    const matchDelta = MATCH_PRIORITY[right.match.kind] - MATCH_PRIORITY[left.match.kind];
    if (matchDelta !== 0) {
      return matchDelta;
    }

    const durationDelta = bestDuration(left) - bestDuration(right);
    if (durationDelta !== 0) {
      return durationDelta;
    }

    return right.listing.dataCompleteness - left.listing.dataCompleteness;
  });
}

function bestDuration(recommendation: Recommendation): number {
  return recommendation.bestRouteMode === "taxi"
    ? recommendation.taxiRoute.durationMinutes
    : recommendation.transitRoute.durationMinutes;
}

function roundMoney(value: number): number {
  return Math.round(value * 100) / 100;
}

File: CarRentalOptimizer/Domain/Recommendation.swift
import Foundation

func buildRecommendation(
    listing: RentalListing,
    match: VehicleMatch,
    taxiRoute: RouteEstimate,
    transitRoute: RouteEstimate
) -> Recommendation {
    let rentalTotal = roundMoney(listing.basePrice + listing.platformFees + listing.insuranceFees + listing.oneWayFee)
    let taxiTotal = roundMoney(rentalTotal + taxiRoute.cost)
    let transitTotal = roundMoney(rentalTotal + transitRoute.cost)
    let bestRouteMode: RouteMode = taxiTotal <= transitTotal ? .taxi : .transit

    return Recommendation(
        listing: listing,
        match: match,
        taxiRoute: taxiRoute,
        transitRoute: transitRoute,
        rentalTotal: rentalTotal,
        taxiTotal: taxiTotal,
        transitTotal: transitTotal,
        bestTotal: bestRouteMode == .taxi ? taxiTotal : transitTotal,
        bestRouteMode: bestRouteMode,
        warnings: Array(Set(listing.warnings))
    )
}

func rankRecommendations(_ recommendations: [Recommendation]) -> [Recommendation] {
    recommendations.sorted { left, right in
        if left.bestTotal != right.bestTotal {
            return left.bestTotal < right.bestTotal
        }
        if left.match.kind != right.match.kind {
            return left.match.kind > right.match.kind
        }
        let leftDuration = bestDuration(left)
        let rightDuration = bestDuration(right)
        if leftDuration != rightDuration {
            return leftDuration < rightDuration
        }
        return left.listing.dataCompleteness > right.listing.dataCompleteness
    }
}

private func bestDuration(_ recommendation: Recommendation) -> Int {
    recommendation.bestRouteMode == .taxi
        ? recommendation.taxiRoute.durationMinutes
        : recommendation.transitRoute.durationMinutes
}

private func roundMoney(_ value: Double) -> Double {
    (value * 100).rounded() / 100
}
